include/envoy/network/transport_socket.h

```diff
@@ -149,9 +151,12 @@ class TransportSocketFactory {
   virtual bool implementsSecureTransport() const PURE;
 
   /**
+   * @param override_server_name set server name, disregard the value the factory was
+   * configured with
    * @return Network::TransportSocketPtr a transport socket to be passed to connection.
    */
-  virtual TransportSocketPtr createTransportSocket() const PURE;
+  virtual TransportSocketPtr
+  createTransportSocket(absl::optional<std::string> override_server_name) const PURE;
```

Comment:
This should be const reference everywhere.

Thinking more, I think we would like some sort of TransportSocketOptions to passed into here, like an interface:

```
class TransportSocketOptions {
public:
  virtual ~TransportSocketOptions();
  virtual absl::optional<string> overrideServerName() const PURE;
  virtual void hashKey(std::vector<uint8_t>& key) const PURE;
}

typedef std::shared_ptr<TransportSocketOptions> TransportSocketOptionsSharedPtr;
```

and pass TransportSocketOptionsSharedPtr around. This will help when we want to have more than `override_server_name`, and you can put this into FilterState as something called `UpstreamTransportSocketOption`.

Reply:
@lizan OK, thank you for the feedback, let me implement it.